snapcraft/storeapi/__init__.py

```diff
@@ -13,9 +13,282 @@
+            # FIXME: Cough, we may want to buffer here (and a progress bar
+            # would be nice) -- vila 2016-04-26
+            f.write(download.content)
+        if self.matching_sha512(download_path, expected_sha512):
+            logger.info('Successfully downloaded {} at {}'.format(
+                name, download_path))
+        else:
+            raise SHAMismatchError(download_path, expected_sha512)
+
+    def matching_sha512(self, path, expected_sha512):
+        if not os.path.exists(path):
+            return False
+
+        file_sum = hashlib.sha512()
+        with open(path, 'rb') as f:
+            for file_chunk in iter(
+                    lambda: f.read(file_sum.block_size * 128), b''):
+                file_sum.update(file_chunk)
+        return expected_sha512 == file_sum.hexdigest()
+
+    # Low level helpers
+
+    def post(self, path, data, headers):
+        url = parse.urljoin(self.root_url, path)
+        response = self.session.post(url, data=data, headers=headers)
+        return response
+
+    def close(self):
+        if self.session is not None:
+            self.session.close()
+        if self.sso.session is not None:
+            self.sso.session.close()
+        if self.updown is not None:
+            self.updown.close()
+
+
+# FIXME: Refactor as a base for all clients using requests.Session() (sca,
+# updown, cpi) with (conf, root_url) to build from -- vila 2016-04-27
+class CPIClient(object):
+    """The Click Package Index knows everything about existing snaps.
+
+    https://wiki.ubuntu.com/AppStore/Interfaces/ClickPackageIndex is the
+    canonical reference.
+    """
+
+    def __init__(self, conf):
+        self.conf = conf
+        self.root_url = os.environ.get('UBUNTU_STORE_SEARCH_ROOT_URL',
+                                       UBUNTU_STORE_SEARCH_ROOT_URL)
+        if self.conf.get('package_access') is None:
+            raise InvalidCredentialsError()
+        self.session = requests.Session()
+
+    def search_package(self, snap_name, channel, arch):
+        headers = {
+            'Accept': 'application/hal+json',
+            'X-Ubuntu-Architecture': arch,
+            'X-Ubuntu-Release': DEFAULT_SERIES,
+            'X-Ubuntu-Device-Channel': channel,
+        }
+        params = {
+            'q': 'package_name:"{}"'.format(snap_name),
+            'fields': 'status,download_url,anon_download_url,download_sha512',
+            'size': 1,
+        }
+        resp = self.get('api/v1/search', headers, params=params)
+        embedded = resp.json().get('_embedded', None)
+        if embedded is None:
+            return None
+        else:
+            return embedded['clickindex:package'][0]
+
+    def get(self, path, headers=None, params=None):
+        if headers is None:
+            headers = {}
+        headers.update({'Authorization':
+                        _macaroon_auth(self.conf, 'package_access')})
+        url = parse.urljoin(self.root_url, path)
```

Comment:
This is weird. You are calling get twice, one for the search where this parse.urljoin is clear. But also for the download, where you have path=https://public.apps.staging.ubuntu.com/download-snap/{...}.snap and root_url=https://search.apps.staging.ubuntu.com/. This works just because urljoin returns the path if it can't join them, but it's not clear at all.
I think you should have two methods.